controller/uninstall_controller.go

```diff
@@ -0,0 +1,437 @@
+)
+
+const (
+	GracePeriod = 90 * time.Second
+)
+
+type UninstallController struct {
+	namespace       string
+	force, firstRun bool
+	ds              *datastore.DataStore
+	stopCh          chan struct{}
+
+	volumeInformerSynced      cache.InformerSynced
+	engineInformerSynced      cache.InformerSynced
+	replicaInformerSynced     cache.InformerSynced
+	engineImageInformerSynced cache.InformerSynced
+	nodeInformerSynced        cache.InformerSynced
+	daemonSetInformerSynced   cache.InformerSynced
+
+	queue workqueue.RateLimitingInterface
+}
+
+func NewUninstallController(
+	namespace string,
+	force bool,
+	ds *datastore.DataStore,
+	stopCh chan struct{},
+	volumeInformer lhinformers.VolumeInformer,
+	engineInformer lhinformers.EngineInformer,
+	replicaInformer lhinformers.ReplicaInformer,
+	engineImageInformer lhinformers.EngineImageInformer,
+	nodeInformer lhinformers.NodeInformer,
+	daemonSetInformer v1beta2.DaemonSetInformer,
+) *UninstallController {
+	c := &UninstallController{
+		namespace:                 namespace,
+		force:                     force,
+		firstRun:                  true,
+		ds:                        ds,
+		stopCh:                    stopCh,
+		volumeInformerSynced:      volumeInformer.Informer().HasSynced,
+		engineInformerSynced:      engineInformer.Informer().HasSynced,
+		replicaInformerSynced:     replicaInformer.Informer().HasSynced,
+		engineImageInformerSynced: engineImageInformer.Informer().HasSynced,
+		nodeInformerSynced:        nodeInformer.Informer().HasSynced,
+		daemonSetInformerSynced:   daemonSetInformer.Informer().HasSynced,
+		queue:                     workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "longhorn-uninstall"),
+	}
+
+	volumeInformer.Informer().AddEventHandler(c.controlleeHandler())
+	engineInformer.Informer().AddEventHandler(c.controlleeHandler())
+	replicaInformer.Informer().AddEventHandler(c.controlleeHandler())
+	engineImageInformer.Informer().AddEventHandler(c.controlleeHandler())
+	nodeInformer.Informer().AddEventHandler(c.controlleeHandler())
+	daemonSetInformer.Informer().AddEventHandler(c.controlleeHandler())
+	return c
+}
+
+func (c *UninstallController) controlleeHandler() cache.ResourceEventHandler {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueueControlleeChange() },
+		UpdateFunc: func(old, cur interface{}) { c.enqueueControlleeChange() },
+		DeleteFunc: func(obj interface{}) { c.enqueueControlleeChange() },
+	}
+}
+
+func (c *UninstallController) enqueueControlleeChange() {
+	c.queue.AddRateLimited("uninstall")
+}
+
+func (c *UninstallController) Run() {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	logrus.Infof("Uninstalling...")
+
+	if !controller.WaitForCacheSync("longhorn uninstall", c.stopCh,
+		c.volumeInformerSynced, c.engineImageInformerSynced,
+		c.nodeInformerSynced, c.daemonSetInformerSynced) {
+
+		logrus.Fatalf("Failed to sync informers")
```

Comment:
Don't use `Fatalf` or `panic` deep in the logic. It's hard to control the exception handle flow this way.

There are others usages in this commit.